### ISSUE TITLE
Feature Request: Replicated Padding

### ISSUE THREAD

**Issue by mharradon** [Theano/Theano#5827]
See, for example: https://github.com/torch/nn/blob/master/doc/convolution.md#nn.VolumetricReplicationPadding

Currently when making convolutional layers the by-far easiest approach is to keep the image size fixed when doing convolution. This is typically achieved by padding the edges of the image with zeros. This has a tendency to introduce edge artifacts into the image. These artifacts can be removed by a well-trained net, but it requires significant effort and fine-parameter tuning (similar to the upscaling issue described in http://distill.pub/2016/deconv-checkerboard/).

This issue becomes worse at high dimensions, as well. For 2D images at edges for 3x3 kernels we have 6/9 pixels inside the boundary. Along edges of 3D volumetric images we have only 12/27 kernel voxels inside the image boundary. In the worst case, at volumetric image corners only 8/27 kernel voxels remain in the image. As neural nets are applied to higher dimension datasets (video, scientific images, etc.) in order to fit images onto graphics cards we'll need to use smaller resolutions, as well - 128x128x128 3D images are each 128 times larger than 2D 128x128 images. So we should expect edge effects to become more important.

One feature request already along these lines is reflection padding: https://github.com/Theano/Theano/issues/5639

I'd also suggest adding replicated padding:

```
1 2 3
4 5 6
7 8 9 

p=1
1 1 2 3 3
1 1 2 3 3
4 4 5 6 6
7 7 8 9 9
7 7 8 9 9

p=2
1 1 1 2 3 3 3
1 1 1 2 3 3 3
1 1 1 2 3 3 3
4 4 4 5 6 6 6
7 7 7 8 9 9 9
7 7 7 8 9 9 9
7 7 7 8 9 9 9
```

One can imagine a number of applications where this would be a preferable "estimate" of the edge values to reflection padding. Currently for 3d images and p=1 I'm using the following:

inp = T.concatenate([inp[:,:,[0],:,:],inp,inp[:,:,[-1],:,:]], axis=2)
inp = T.concatenate([inp[:,:,:,[0],:],inp,inp[:,:,:,[-1],:]], axis=3)
inp = T.concatenate([inp[:,:,:,:,[0]],inp,inp[:,:,:,:,[-1]]], axis=4)

Which I believe adds a significant overhead both in compilation and in computation.

The primary use for this would be in convolution with pad='valid', so additional arguments (pad='reflect', pad='repeat') could alternately be added there, especially if it resulted in improved runtime or compile performance.

I wasn't able to find this specific request anywhere else, though some are similar or include it, ex: https://github.com/Theano/Theano/issues/5471

Thanks!

**Comment by nouiz** [comment]
Why do you think repeat would be better then reflect?

Both of them could be done by the img2col operator for the forward operation inside the [Cpu]CorrMM* implementation. That would be more efficient then doing it in the graph. im2col already explicitly add the padding with 0s, it could be modified to do reflect or repeat.

But I don't think it will make a big computation speed difference. If you already use the graph implementation, you can run your graph with theano profiler and give us the full output? (mostly use Theano flag profile=True http://deeplearning.net/software/theano/tutorial/profiling.html)

If someone want to implement it, the user interface to replace valid with reflect or repeat don't seem to work I think. The problem is that it don't allow to tell by how many it pad. Maybe we need a new parameter like pad_type=[0,reflect,repeat].

**Comment by mharradon** [comment]
> Why do you think repeat would be better then reflect?

I think it will vary by use case, but you can imagine many images with non-homogenous structure - bright, or smooth in one corner and not in another corner. In that case reflection would introduce irrelevant or incorrect features into the convolution. On the other hand, reflection could better maintain the spatial structure of features, whereas repetition would induce a highly structured spatial pattern. I think for most applications I would consider them both worth testing.

> But I don't think it will make a big computation speed difference...

You may be right, but my understanding is that it would be a major compilation overhead in terms of adding nodes for each convolution.

>.. you can run your graph with theano profiler and give us the full output? (mostly use Theano flag profile=True http://deeplearning.net/software/theano/tutorial/profiling.html)

Will do, coming shortly.

>If someone want to implement it, the user interface to replace valid with reflect or repeat don't seem to work I think. The problem is that it don't allow to tell by how many it pad. Maybe we need a new parameter like pad_type=[0,reflect,repeat].

Yes, good point. That would be a great solution for my use cases.

**Comment by nouiz** [comment]
We have done much work on the graph optimization time. If you see a big
difference of using it vs not, tell us. There should be some slowdown, but
probably not too much.
